src/FsAutoComplete/CodeFixes/RemoveUnnecessaryParentheses.fs

```
@@ -0,0 +1,168 @@
+              match twoBefore, oneBefore, s[0] with
+              | _, _, ('\n' | '\r') -> None
+              | '[', '|', (Punctuation | LetterOrDigit) -> None
+              | _, '[', '<' -> Some ShouldPutSpaceBefore
+              | _, ('(' | '[' | '{'), _ -> None
+              | _, '>', _ -> Some ShouldPutSpaceBefore
+              | ' ', '=', _ -> Some ShouldPutSpaceBefore
+              | _, '=', ('(' | '[' | '{') -> None
+              | _, '=', (Punctuation | Symbol) -> Some ShouldPutSpaceBefore
+              | _, LetterOrDigit, '(' -> None
+              | _, (LetterOrDigit | '`'), _ -> Some ShouldPutSpaceBefore
+              | _, (Punctuation | Symbol), (Punctuation | Symbol) -> Some ShouldPutSpaceBefore
+              | _ -> None)
+            |> Option.flatten
+
+          let (|ShouldPutSpaceAfter|_|) (s: string) =
+            // (……)…
+            //   ↑ ↑
+            sourceText.TryGetChar(range.End.IncColumn 1)
+            |> Option.bind (fun endChar ->
+              match s[s.Length - 1], endChar with
+              | '>', ('|' | ']') -> Some ShouldPutSpaceAfter
+              | _, (')' | ']' | '[' | '}' | '.' | ';' | ',' | '|') -> None
+              | (Punctuation | Symbol), (Punctuation | Symbol | LetterOrDigit) -> Some ShouldPutSpaceAfter
+              | LetterOrDigit, LetterOrDigit -> Some ShouldPutSpaceAfter
+              | _ -> None)
+
+          match adjusted with
+          | ShouldPutSpaceBefore & ShouldPutSpaceAfter -> " " + adjusted + " "
+          | ShouldPutSpaceBefore -> " " + adjusted
+          | ShouldPutSpaceAfter -> adjusted + " "
+          | adjusted -> adjusted
+
+        return
+          [ { Edits = [| { Range = d.Range; NewText = newText } |]
+              File = codeActionParams.TextDocument
+              Title = title
+              SourceDiagnostic = Some d
+              Kind = FixKind.Fix } ]
+
+      | notParens ->
+        System.Diagnostics.Debug.Fail $"%A{notParens} <> ('(', ')')"
```

can you log something here instead of failing hard?

---

Hmm, the only way it would really be hit at this point would be if this fix were triggered for the wrong diagnostic ID, so maybe I'll just take it out.